client/ayon_silhouette/api/lib.py

```diff
@@ -0,0 +1,243 @@
+    #     fps_def = NumberDef(
+    #         "fps", label="FPS", default=current_fps, decimals=5
+    #     )
+    #     defs.append(fps_def)
+
+    return defs
+
+
+@contextlib.contextmanager
+def maintained_selection():
+    """Maintain selection during context."""
+
+    previous_selection = fx.selection()
+    try:
+        yield
+    finally:
+        fx.select(previous_selection)
+        pass
+
+
+@contextlib.contextmanager
+def undo_chunk(label=""):
+    """Open undo chunk during context.
+
+    In Silhouette, it's often necessary to group operations into an undo chunk
+    to ensure the UI updates correctly on property and value changes.
+
+    Note that `contextlib.contextmanager` can also be used as function
+    decorators.
+
+    """
+    try:
+        fx.beginUndo(label)
+        yield
+    finally:
+        fx.endUndo()
+
+
+def imprint(node, data: Optional[dict], key="AYON"):
+    """Write `data` to `node` as userDefined attributes
+
+    Arguments:
+        node (fx.Object | fx.Node): The selection object
+        data (dict): Dictionary of key/value pairs
+    """
+
+    # Remove data
+    if data is None:
+        if isinstance(node, fx.Node):
+            node.setState(key, None)
+        else:
+            node.removeProperty(key)
+        return
+
+    # Set data
+    if isinstance(node, fx.Node):
+        node.setState(key, data)
+    elif isinstance(node, fx.Object):
+        # Serialize data
+        value = json.dumps(data)
+        prop = node.property(key)
+        if not prop:
+            # Create property
+            # Arguments are `id`, `label` and `default_value`. By passing the
+            # default value as empty string we make it a string attribute.
+            prop = fx.Property(key, key, "")
+            prop.hidden = True
+            node.addProperty(prop)
+
+        # Set value
+        prop.value = value
+    else:
+        raise TypeError(f"Unsupported node type: {node} ({type(node)})")
+
+
+def read(node, key="AYON") -> Optional[dict]:
+    """Return user-defined attributes from `node`
+
+    Arguments:
+        node (fx.Object | fx.Node): Node or object to redad from.
+        key (str): The key to read from.
+
+    Returns:
+        Optional[dict]: The data stored in the node.
+
+    """
+    if isinstance(node, fx.Node):
+        # Use node state instead of property
+        return node.getState(key)
+    elif isinstance(node, fx.Object):
+        # Project or source items do not have state
+        prop = node.property(key)
+        if not prop:
+            return
+        value = prop.value
+        if not value:
+            return
+
+        try:
+            return json.loads(value)
+        except json.JSONDecodeError as exc:
+            log.error(
+                f"Failed to read '{key}' from node {node}"
+                f" with value {value}: {exc}")
+            return
+    else:
+        raise TypeError(f"Unsupported node type: {node} ({type(node)})")
```

Comment:
```suggestion
    raise TypeError(f"Unsupported node type: {node} ({type(node)})")
```

Reply:
Maybe it would worth to reverse the condition at the top to first raise error if is not `fx.Object`.

Reply:
I'm not sure `fx.Node` inherits from `fx.Object` though :') but will check.

Reply:
```python
 if isinstance(node, fx.Node):
    # Use node state instead of property
    return node.getState(key)

if not isinstance(node, fx.Object):
    raise TypeError(f"Unsupported node type: {node} ({type(node)})")

...
```

Reply:
I like that less because if I were to add another class type I'd need to restructure the whole thing regardless.

I prefer this particular case the way it is now - it's much clearer that there is just these three branches.